### PR TITLE
Adds Dark Theme Mode

### LINKED ISSUE #530: Add Dark Theme to SCE Website
### How to do this
1. Add a [`CustomInput`](https://reactstrap.github.io/components/form/#Custom-Inputs) react component to [`src/Components/Navbar/UserNavbar.js`](https://github.com/SCE-Development/Core-v4/blob/dev/src/Components/Navbar/UserNavbar.js).
1. The `CustomInput` should have an `onClick` handler that sets a cookie to `true` or `false` within [`src/Components/Navbar/UserNavbar.js`](https://github.com/SCE-Development/Core-v4/blob/dev/src/Components/Navbar/UserNavbar.js).
1. We now need to handle whether the cookie is `true` or `false`. We can handle the logic for the `body` class name in [`src/index.js`](https://github.com/SCE-Development/Core-v4/blob/dev/src/index.js). Add a function called `handleDarkTheme` which checks if the cookie's value and modifies the body class name as a result.
1. Call the class from the `useEffect` in [`src/index.js`](https://github.com/SCE-Development/Core-v4/blob/dev/src/index.js).

### PR DESCRIPTION
**Pull request by fmasood23**
Resolves #530

### CODE REVIEW

File: src/Components/Navbar/UserNavbar.js [hunk 1]
```diff
@@ -76,11 +97,45 @@ export default function UserNavBar(props) {
...
+      cook.set('thisCookie', 'true', { path: '/' });
+    } else {
+      document.body.className = 'light';
+      setDarkTheme(false);
+      disableDarkMode();
+      cook.set('thisCookie', 'false', { path: '/' });
+    }
+  }
+
+
```
Comment (seemavora) [inline src/Components/Navbar/UserNavbar.js:120]: remove blank line

File: src/Components/Navbar/UserNavbar.js [hunk 2]
```diff
@@ -89,13 +144,15 @@ export default function UserNavBar(props) {
...
         </NavbarBrand>
+        <CustomInput onChange={createCookie} className="darkToggle" type="switch" id="exampleCustomSwitch" name="customSwitch" checked={darkTheme} />
         <NavbarToggler tag='h1'>
           <ButtonDropdown isOpen={menuIsOpen} toggle={toggler}>
             <DropdownToggle className='hamburger-button'>
               <span className='border'></span>
               <span className='border'></span>
               <span className='border'></span>
             </DropdownToggle>
+
```
Comment (seemavora) [inline src/Components/Navbar/UserNavbar.js:155]: remove blank line

File: src/index.css [hunk 3]
```diff
@@ -17,3 +17,82 @@ code {
...
+}
+
+body.dark.vimvixen-hint {
+  background-color: #7b5100 !important;
+  border-color: #dcb20d !important;
+  color: #f9edcc !important;
+}
+body.dark::placeholder {
+  opacity: 0.5 !important;
+}
```
Comment (seemavora) [inline src/index.css:98]: need a newline at the end of files

File: src/index.js [hunk 4]
```diff
@@ -37,3 +70,5 @@ function App(props) {
 export default withRouter(App);
 
 ReactDOM.render(<App />, document.getElementById('root'));
+
+
```
Comment (seemavora) [inline src/index.js:74]: remove extra blank line

**Comment by fmasood23** [comment]
So basically dark reader was something I imported- maybe the package.json didn’t get updated let me check real quick

**Comment by evanugarte** [comment]
@fmasood23 do `npm install darkreader --save` in the topmost directory in Core-v4. after commit and push the changes to the `package` files

**Comment by fmasood23** [comment]
Made changes to the formatting (extra lines to delete and add) and so you can access the darkreader and universal cookie - let me know if this doesn't work!

**Comment by fmasood23** [comment]
@ seema 
so since the dark mode is from an installed package the colors are already set but i will look into it and see if there is a way to make those icons lighter

File: src/Components/Navbar/UserNavbar.js [hunk 5]
```diff
@@ -15,9 +17,29 @@ import {
...
 import { membershipState } from '../../Enums';
 import logo from '../Navbar/sce_logo.png';
+import {
+  enable as enableDarkMode,
+  disable as disableDarkMode,
+  auto as followSystemColorScheme,
+  disable,
+} from 'darkreader';
+
+enableDarkMode({
```
Comment (evanugarte) [inline src/Components/Navbar/UserNavbar.js:27]: do you need lines `27-33`?

File: src/Components/Navbar/UserNavbar.js [hunk 6]
```diff
@@ -76,11 +98,37 @@ export default function UserNavBar(props) {
...
+  function keepState() {
+    if (cook.get('thisCookie') === 'true') {
+      setDarkTheme(true);
+    } else {
+      setDarkTheme(false);
+    }
+  }
+
+  useEffect(() => {
+    keepState();
```
Comment (evanugarte) [inline src/Components/Navbar/UserNavbar.js:128]: perhaps rename this to
```suggestion
    checkIfDarkThemeActive();
```

File: src/Components/Navbar/UserNavbar.js [hunk 7]
```diff
@@ -111,17 +162,17 @@ export default function UserNavBar(props) {
                 </NavItem>
               </DropdownItem>
               {props.user && props.user.accessLevel >=
-              membershipState.MEMBER && (
+                membershipState.MEMBER && (
```
Comment (evanugarte) [inline src/Components/Navbar/UserNavbar.js:165]: hmm looks like prettier changed a bunch of lines unecesarrily. id move these indented lines back to where they were and double check style with `npm run lint`

File: src/index.css [hunk 8]
```diff
@@ -17,3 +17,83 @@ code {
   font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
     monospace;
 }
+
+/* body.dark html {
```
Comment (evanugarte) [inline src/index.css:21]: remove this commented out stuff if were not using it

File: src/index.js [hunk 9]
```diff
@@ -2,10 +2,29 @@ import React, { useEffect, useState } from 'react';
...
+import { setFetchMethod } from 'darkreader';
+
+import {
+  enable as enableDarkMode,
+  disable as disableDarkMode,
+  auto as followSystemColorScheme,
+} from 'darkreader';
+
+setFetchMethod(window.fetch);
+enableDarkMode({
```
Comment (evanugarte) [inline src/index.js:21]: do we need `21-27`?

File: src/index.js [hunk 10]
```diff
@@ -20,8 +39,19 @@ function App(props) {
...
+      enableDarkMode();
+    } else {
+      document.body.className = 'light';
+      disableDarkMode();
+    }
+  }
+
   useEffect(() => {
     getAuthStatus();
+    handleDarkTheme();
```
Comment (evanugarte) [inline src/index.js:54]: i see that we are doing a similar approach in the navbar as well. i suggest creating a file in [`src/APIFunctions`](https://github.com/SCE-Development/Core-v4/tree/dev/src/APIFunctions) called `darkTheme.js`. There u can define two functions for toggling and checking if the darktheme is active

```js
// put all the imports u need here

export function checkIfDarkThemeActive() {
  // paste ur code for checking
}

export function toggleDarkTheme() {
  // paste ur code for toggling
}
```

after u can just import these functions at the top of index.js and UserNavbar.js and call on them without duplicate code.

File: src/Components/Navbar/UserNavbar.js (hunk 7 above)
Comment (fmasood23) [reply src/Components/Navbar/UserNavbar.js:165]: When I ran `npm run lint ` everything came back okay with the current formatting.

File: src/index.js (hunk 9 above)
Comment (fmasood23) [reply src/index.js:21]: 21 - 27 is part of the import for the darkreader

File: src/Components/Navbar/UserNavbar.js (hunk 7 above)
Comment (seemavora) [reply src/Components/Navbar/UserNavbar.js:165]: if u ever did the mosh tutorial and enabled one of the vscode extentions he asked u to install, u will have to uninstall it bc it messes with the way we structure/align our code

Comment (fmasood23) [reply src/Components/Navbar/UserNavbar.js:165]: i uninstalled it!

Comment (seemavora) [reply src/Components/Navbar/UserNavbar.js:165]: try saving the file, and quitting out of vscode and opening it up again, if the indents are still weird u may have to manually fix some of the changes this time around

**Comment by seemavora** [comment]
Travis failed at two tests, you can see exactly where right here: https://travis-ci.com/github/SCE-Development/Core-v4/jobs/359549727

File: src/index.js (hunk 10 above)
Comment (evanugarte) [reply src/index.js:54]: 1. you wanna call on `toggleDarkTheme` whenever u click on the switch.
1. in `index.js`, call on `checkIfDarkThemeActive` from `useEffect`

the navbar and index.js should contain no other dark theme logic, only the api functions. the api functions should take care of everything

**Comment by nanarboursal** [comment]
Screen Shot 2020-07-24 at 9 29 55 AM

**Comment by fmasood23** [comment]
I ran `npm install darkreader --save` after committing and i pushed my changes but the darkreader still doesn't work?

**Comment by evanugarte** [comment]
@nanarboursal @seemavora did yall do `npm install` before doing `npm run frontend`?

**Comment by nanarboursal** [comment]
works, but don't forget to update the branch, and the travis tests still die!

**Comment by seemavora** [comment]
Works! Do we need the the small settings img in the navbar tho? i think just the moon looks fine. 
image

**Comment by fmasood23** [comment]
LMAO the "settings" one was supposed to be a sun - i can change the icon to look more like a sun? or would you suggest getting rid of it entirely?

**Comment by seemavora** [comment]
> lol the settings one was supposed to be a sun - i can change the icon to another one that looks more like a sun? or would you suggest getting rid of it entirely?

oh my bad! try playing around with it. The moon and sun kinda take attention away from the logo in my opinion, but everything else looks good!

**Comment by fmasood23** [comment]
This is what i can change it to look like - lmk if you think it is too distracting. If we want to keep the icons, we can move the entire thing to the other side of the navbar, next to the social media icons? let me know what you think

Screen Shot 2020-07-24 at 12 05 33 PM
Screen Shot 2020-07-24 at 12 05 43 PM

**Comment by fmasood23** [comment]
made the icons a little smaller to be less distracting. let me know what you think

Screen Shot 2020-07-24 at 12 10 31 PM
Screen Shot 2020-07-24 at 12 10 22 PM

**Comment by qsaml** [comment]
curious question, have you tried putting the darkmode in the member dropdown? I like the way it looks!

**Comment by seemavora** [comment]
Looks better! How about using something like this for the sun? Bc at first glance it doesn't look like the sun hahaha 
image

**Comment by fmasood23** [comment]
will look for an icon that looks like that! also tried moving the whole toggle to the other end of the navbar (in the dropdown) and it looked pretty odd so that isn't an option

**Comment by qsaml** [comment]
ah I see. again still looks really nice

**Comment by fmasood23** [comment]
@seemavora changed the icons to fit your suggestion - what do you think?

Screen Shot 2020-07-24 at 12 32 20 PM
Screen Shot 2020-07-24 at 12 32 28 PM

**Comment by fmasood23** [comment]
> ah I see. again still looks really nice

thanks!

**Comment by seemavora** [comment]
looks much better! make sure you update your branch and also fix travis. :))

**Comment by fmasood23** [comment]
> looks much better! make sure you update your branch and also fix travis. :))

yes will work on that thank you!

**Comment by hsorathia** [comment]
looks hella sick, good work yo

**Comment by qsaml** [comment]
very nice and well done!

**Comment by fmasood23** [comment]
thank you guys!

**Pull request merged**